chips/msp432/src/adc.rs

```diff
@@ -0,0 +1,982 @@
+            Mem19 = 0x32,
+            /// MEM20 interrupt flag
+            Mem20 = 0x34,
+            /// MEM21 interrupt flag
+            Mem21 = 0x36,
+            /// MEM22 interrupt flag
+            Mem22 = 0x38,
+            /// MEM23 interrupt flag
+            Mem23 = 0x3A,
+            /// MEM24 interrupt flag
+            Mem24 = 0x3C,
+            /// MEM25 interrupt flag
+            Mem25 = 0x3E,
+            /// MEM26 interrupt flag
+            Mem26 = 0x40,
+            /// MEM27 interrupt flag
+            Mem27 = 0x42,
+            /// MEM28 interrupt flag
+            Mem28 = 0x44,
+            /// MEM29 interrupt flag
+            Mem29 = 0x46,
+            /// MEM30 interrupt flag
+            Mem30 = 0x48,
+            /// MEM31 interrupt flag
+            Mem31 = 0x4A,
+            /// RDYIFG interrupt flag
+            Ready = 0x4C
+        ]
+    ]
+];
+
+/// Create a trait of both client types to allow a single client reference to
+/// act as both
+pub trait EverythingClient: hil::adc::Client + hil::adc::HighSpeedClient {}
+impl<C: hil::adc::Client + hil::adc::HighSpeedClient> EverythingClient for C {}
+
+pub struct Adc {
+    registers: StaticRef<AdcRegisters>,
+    resolution: AdcResolution,
+    mode: Cell<AdcMode>,
+    active_channel: Cell<Channel>,
+    ref_module: OptionalCell<&'static dyn ref_module::AnalogReference>,
+    timer: OptionalCell<&'static dyn timer::InternalTimer>,
+    dma: OptionalCell<&'static dma::DmaChannel<'static>>,
+    pub(crate) dma_chan: usize,
+    dma_src: u8,
+    buffer1: TakeCell<'static, [u16]>,
+    buffer2: TakeCell<'static, [u16]>,
+    client: OptionalCell<&'static dyn EverythingClient>,
+}
+
+#[repr(u32)]
+#[derive(Copy, Clone, PartialEq)]
+pub enum Channel {
+    Channel0 = 0,
+    Channel1 = 1,
+    Channel2 = 2,
+    Channel3 = 3,
+    Channel4 = 4,
+    Channel5 = 5,
+    Channel6 = 6,
+    Channel7 = 7,
+    Channel8 = 8,
+    Channel9 = 9,
+    Channel10 = 10,
+    Channel11 = 11,
+    Channel12 = 12,
+    Channel13 = 13,
+    Channel14 = 14,
+    Channel15 = 15,
+    Channel16 = 16,
+    Channel17 = 17,
+    Channel18 = 18,
+    Channel19 = 19,
+    Channel20 = 20,
+    Channel21 = 21,
+    Channel22 = 22,
+    Channel23 = 23,
+}
+
+#[allow(dead_code)]
+#[repr(u32)]
+#[derive(Copy, Clone, PartialEq)]
+enum AdcResolution {
+    Bits8 = 0,
+    Bits10 = 1,
+    Bits12 = 2,
+    Bits14 = 3,
+}
+
+#[allow(dead_code)]
+#[derive(Copy, Clone, PartialEq)]
+enum AdcMode {
+    Single,
+    Repeated,
+    Highspeed,
+    Disabled,
+}
+
+fn buf_u8_to_buf_u16(buf: &'static mut [u8]) -> &'static mut [u16] {
+    let buf_ptr = unsafe { mem::transmute::<*mut u8, *mut u16>(buf.as_mut_ptr()) };
+    unsafe { slice::from_raw_parts_mut(buf_ptr, buf.len() / 2) }
+}
+
+fn buf_u16_to_buf_u8(buf: &'static mut [u16]) -> &'static mut [u8] {
+    let buf_ptr = unsafe { mem::transmute::<*mut u16, *mut u8>(buf.as_mut_ptr()) };
+    unsafe { slice::from_raw_parts_mut(buf_ptr, buf.len() * 2) }
+}
```

Comment:
These functions are not safe to call for arbitrary input slices, and so should be marked `unsafe` and their requirements documented. If an odd length slice is passed to `buf_u8_to_buf_u16`, it should probably panic rather than go ahead with a transmute. 

This is chip specific code, so we don't really have to worry about endianness, but we do have to consider the possibility for alignment issues when converting from u8 to u16 -- the individual bytes might not be aligned for access as u16s.

Reply:
Ahh yeah, of course! I completely oversaw that fact, that u8 to u16 can be a big problem.
I'll fix that as soon as possible!

Reply:
With the latest commit I marked the 2 functions as unsafe and documented why they are necessary and what the requirement for `buf_u8_to_buf_u16` is. I hope these functions now correctly implemented.